### PR TITLE
Remove opt to generate llvmbc from llvmir and clean up spirv_generator.py

### LINKED ISSUE #249: Remove the explicit call to opt inside spirv-generator
Right now we have a call inside spirv_generator.py https://github.com/IntelPython/numba-dppy/blob/b3eff86e6ba4f4b2c0795d09f0d57aa14ed18883/numba_dppy/spirv_generator.py#L85 where we invoke `opt` this step should be removed and we should directly pass the binary IR to `llvm-spirv`.

### PR DESCRIPTION
**Pull request by reazulhoque**
Closes #249 

This PR:
1. Adds documentation in spirv_generator.py.
2. Cleans up spirv_generator.py.
3. Dumps LLVM IR, LLVM Bitcode and generated SPIRV in the directory the program was run from when users provide `NUMBA_DPPY_SAVE_IR_FILES`.
4. Removes usage of `opt` to generate llvmbc from llvmir

### CODE REVIEW

File: numba_dppy/spirv_generator.py [hunk 1]
```diff
@@ -34,91 +34,103 @@ def _raise_bad_env_path(msg, path, extra=None):
...
+
+        We use llvm-spirv tool to translate a llvm bitcode into spirv.
 
+        Args:
+            llvm_spirv_args: Args to be provided to llvm-spirv tool.
+            ipath: Input file path of the llvm bitcode.
+            opath: Output file path of the generated spirv.
+        """
         llvm_spirv_flags = []
         if config.DEBUG:
```
Comment (diptorupd) [inline numba_dppy/spirv_generator.py:102]: If this flag is not being used anywhere, then can the code be removed?

File: numba_dppy/spirv_generator.py [hunk 2]
```diff
@@ -34,91 +34,103 @@ def _raise_bad_env_path(msg, path, extra=None):
...
+        """
+        Link spirv modules.
 
-        check_call(params)
+        Args:
+            opath: Output file path of the linked final spirv.
+            binaries: Spirv modules to be linked.
+        """
+        flags = ["--allow-partial-linkage"]
+        check_call(["spirv-link", *flags, "-o", opath, *binaries])
```
Comment (diptorupd) [inline numba_dppy/spirv_generator.py:129]: `spirv-link` is deprecated and we should use `llvm-link`.

File: numba_dppy/spirv_generator.py [hunk 3]
```diff
@@ -221,7 +254,7 @@ def finalize(self):
 # Public llvm_to_spirv function ###############################################
```
Comment (diptorupd) [inline numba_dppy/spirv_generator.py:254]: remove this comment.

**Comment by diptorupd** [comment]
Convert the comment at the top of the file into a module docstring.

File: numba_dppy/spirv_generator.py [hunk 4]
```diff
@@ -170,24 +164,46 @@ def finalize(self):
...
-        llvm_spirv_args = None
+
+        llvm_spirv_args = []
         for key in list(self.context.extra_compile_options.keys()):
             if key == LINK_ATOMIC:
                 from .ocl.atomics import get_atomic_spirv_path
 
                 binary_paths.append(get_atomic_spirv_path())
-            if key == LLVM_SPIRV_ARGS:
+            elif key == LLVM_SPIRV_ARGS:
```
Comment (diptorupd) [inline numba_dppy/spirv_generator.py:174]: Are these two options mutually exclusive?

Comment (reazulhoque) [reply numba_dppy/spirv_generator.py:174]: No, but key will never be both.

**Review by diptorupd — APPROVED** [review]

**Pull request merged**